### PR TITLE
Add docs generation command

### LINKED ISSUE #1537: Generate flags documentation
Cobra allows generating documentation based on commands and flags https://github.com/spf13/cobra#generating-documentation-for-your-command.

This documentation could be made available on our documentation. People often ask what options are we supporting. The operator could also leverage these docs as flags are passed directly into options map in CR.

cc @objectiser

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #1537 

docs PR: https://github.com/jaegertracing/documentation/pull/257

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

File: cmd/docs/command.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+func Command() *cobra.Command {
+	return &cobra.Command{
+		Use:   "docs",
+		Short: "Generates documentation",
+		Long:  `Generates command and flags documentation`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			for cmd.Parent() != nil {
+				cmd = cmd.Parent()
+			}
+			return doc.GenMarkdownTree(cmd, "./")
```
Comment (yurishkuro) [inline cmd/docs/command.go]: ran this locally, it's kind of misleading that there's no output from the command. I would recommend at least some log file "generating .md files in the current directory"

File: cmd/docs/docs.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+## docs
```
Comment (yurishkuro) [inline cmd/docs/docs.md]: does this file need to be checked in?

File: Gopkg.toml [hunk 3]
```diff
@@ -30,7 +30,8 @@ required = [
     "github.com/golang/protobuf/protoc-gen-go",
     "github.com/gogo/protobuf/protoc-gen-gogo",
     "github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway",
-    "github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger"
+    "github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger",
+    "github.com/russross/blackfriday"
```
Comment (yurishkuro) [inline Gopkg.toml]: why do we need it here?

File: Gopkg.toml [hunk 4]
```diff
@@ -49,6 +50,11 @@ required = [
   name = "github.com/bsm/sarama-cluster"
   version = "=2.1.13"
 
+[[constraint]]
+  name = "github.com/russross/blackfriday"
+  version = "=1.5"
```
Comment (yurishkuro) [inline Gopkg.toml]: Why do we need this as explicit dependency, instead of transitive via cobra?

File: cmd/all-in-one/main.go [hunk 5]
```diff
@@ -165,6 +167,11 @@ func main() {
 		strategyStoreFactory.AddFlags,
 	)
 
+	//err = doc.GenMarkdownTree(command, "./")
+	//if err != nil {
+	//	log.Fatal(err)
+	//}
```
Comment (yurishkuro) [inline cmd/all-in-one/main.go]: plz remove

**Comment by pavolloffay** [comment]
Cobra can also output html and man pages. We could even make it configurable via a flag.

Most importantly we should go with a format which can be embedded to our docs. I am not sure if we can easily embed there htlm. The markdown can be easily embedded see the PR to the website.

File: Gopkg.toml (hunk 4 above)
Comment (pavolloffay) [reply Gopkg.toml]: I could not make it work that way

File: Gopkg.toml (hunk 3 above)
Comment (pavolloffay) [reply Gopkg.toml]: It's needed by cobra and the automatic transitive resolution didn't work.

**Comment by pavolloffay** [comment]
> actually, I don't think pre-formatted output in .md file is the best format. I see that blackfriday supports HTML output, not sure if it makes a different and/or of cobra can generate HTML.

actually cobra cannot generate html, I have added a flag to generate output in md, rst and man format.

File: cmd/docs/command.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+				cmd = cmd.Parent()
+			}
+			log.Println("Generating documentation in the current working directory")
+			switch v.GetString(formatFlag) {
+			case "md":
+				doc.GenMarkdownTree(cmd, "./")
+			case "man":
+				return man(cmd)
+			case "rst":
+				doc.GenReSTTree(cmd, "./")
```
Comment (yurishkuro) [inline cmd/docs/command.go]: Why do some of these return error and others don't?

File: cmd/docs/command.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		Short: "Generates documentation",
+		Long:  `Generates command and flags documentation`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			for cmd.Parent() != nil {
+				cmd = cmd.Parent()
+			}
+			log.Println("Generating documentation in the current working directory")
+			switch v.GetString(formatFlag) {
+			case "md":
+				doc.GenMarkdownTree(cmd, "./")
```
Comment (yurishkuro) [inline cmd/docs/command.go]: Should it also take the output dir param? You gitignored cmd/docs, which seems different.

File: cmd/docs/command.go (hunk 6 above)
Comment (pavolloffay) [reply cmd/docs/command.go]: +1 done

**Comment by pavolloffay** [comment]
@yurishkuro one last thing. Do we like more `jaeger docs` or `jaeger gen docs`. The later is more flexible if we add more gen commands (e.g. auto-completion).

File: cmd/agent/main.go [hunk 8]
```diff
@@ -84,6 +85,7 @@ func main() {
 	}
 
 	command.AddCommand(version.Command())
+	command.AddCommand(docs.Command())
```
Comment (yurishkuro) [inline cmd/agent/main.go]: missing `v` arg?

File: cmd/docs/command.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+func Command(v *viper.Viper) *cobra.Command {
+	c := &cobra.Command{
+		Use:   "docs",
+		Short: "Generates documentation",
+		Long:  `Generates command and flags documentation`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			for cmd.Parent() != nil {
+				cmd = cmd.Parent()
+			}
+			log.Println("Generating documentation in the current working directory")
```
Comment (yurishkuro) [inline cmd/docs/command.go]: `in ${dir}`, not current dir

File: cmd/docs/command.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+			dir := v.GetString(dirFlag)
+			switch v.GetString(formatFlag) {
+			case "md":
+				return doc.GenMarkdownTree(cmd, dir)
+			case "man":
+				return man(cmd, dir)
+			case "rst":
+				return doc.GenReSTTree(cmd, dir)
```
Comment (yurishkuro) [inline cmd/docs/command.go]: let's add
```go
			case "yaml":
				return doc.GenYamlTree(cmd, dir)
```

**Pull request merged**